ludwig/config_validation/checks.py

```diff
@@ -0,0 +1,327 @@
+            return
+        if TYPE not in config[COMBINER]:
+            raise ConfigValidationError("Combiner must have a type.")
+        if config[COMBINER][TYPE] not in get_combiner_registry():
+            raise ConfigValidationError(f"Combiner type must be one of: {list(get_combiner_registry().keys())}.")
+
+    # Check trainer.
+    if TRAINER in config and config[TRAINER] is None:
+        raise ConfigValidationError("Trainer cannot be None.")
+
+    # Check optimizer.
+    if TRAINER in config and "optimizer" in config[TRAINER]:
+        if config[TRAINER]["optimizer"] is None:
+            raise ConfigValidationError("Trainer.optimizer cannot be None.")
+        if TYPE in config[TRAINER]["optimizer"]:
+            if config[TRAINER]["optimizer"][TYPE] not in optimizer_registry:
+                raise ConfigValidationError(
+                    f"Trainer.optimizer.type must be one of: {list(optimizer_registry.keys())}."
+                )
+
+
+@register_config_check
+def check_feature_names_unique(config: ModelConfigDict) -> None:
+    """Checks that all feature names are unique."""
+    input_features = config[INPUT_FEATURES]
+    input_feature_names = {input_feature[NAME] for input_feature in input_features}
+
+    output_features = config[OUTPUT_FEATURES]
+    output_feature_names = {output_feature[NAME] for output_feature in output_features}
+
+    if len(input_feature_names) + len(output_feature_names) != len(input_features) + len(output_features):
+        raise ConfigValidationError("Feature names must be unique.")
+
+
+@register_config_check
+def check_tied_features_valid(config: ModelConfigDict) -> None:
+    """Checks that all tied features are valid."""
+    print("CHECKING TIED FEATURES!")
+    input_features = config[INPUT_FEATURES]
+    input_feature_names = {input_feature[NAME] for input_feature in input_features}
+
+    for input_feature in input_features:
+        if input_feature["tied"] and input_feature["tied"] not in input_feature_names:
+            raise ConfigValidationError(
+                f"Feature {input_feature[NAME]} is tied to feature {input_feature['tied']}, but the "
+                f"'{input_feature['tied']}' feature does not exist."
+            )
+
+
+@register_config_check
+def check_training_runway(config: ModelConfigDict) -> None:
+    """Checks that checkpoints_per_epoch and steps_per_checkpoint aren't simultaneously defined."""
+    if config[MODEL_TYPE] == MODEL_ECD:
+        if config[TRAINER]["checkpoints_per_epoch"] != 0 and config[TRAINER]["steps_per_checkpoint"] != 0:
+            raise ConfigValidationError(
+                "It is invalid to specify both trainer.checkpoints_per_epoch AND "
+                "trainer.steps_per_checkpoint. Please specify one or the other, or specify neither to "
+                "checkpoint/eval the model every epoch."
+            )
+
+
+@register_config_check
+def check_gbm_horovod_incompatibility(config: ModelConfigDict) -> None:
```

Comment:
This is fine for now because we don't validate on the backend, but can be removed in the future since we have separate schemas for ECD and GBM.